### DISCUSSION TITLE
How to perform OTA operations on ESP32 target with Zephyr 3.5.0 ?

### DISCUSSION THREAD

**Discussion by joelguittet** [zephyrproject-rtos/zephyr#66457] [Q&A]
Hello community!

I would like to perform OTA on ESP32 with Zephyr.

I have build the hello world example with the following command:
```
west build -b esp32_devkitc_wroom --sysbuild samples/hello_world
```

This is generating mcuboot and application. After flashing with `west flash` and starting espressif monitor with `west espressif monitor` I have the following result:

```
Serial port /dev/ttyUSB0
Connecting.....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting....
Detecting chip type... ESP32
--- idf_monitor on /dev/ttyUSB0 115200 ---
--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
ets Jun  8 2016 00:22:57
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0000,len:6764
ho 0 tail 12 room 4
load:0x40078000,len:7112
ho 0 tail 12 room 4
load:0x4009c000,len:1024
load:0x4009c400,len:25132
entry 0x4009dd80
[esp32] [INF] MCUboot 2nd stage bootloader
[esp32] [INF] compiled at Dec 12 2023 22:40:27
[esp32] [INF] Chip revision: 1
[esp32] [INF] Chip revision: 1, min. bootloader chip revision: 0
[esp32] [INF] SPI Flash speed: 40MHz, mode: DIO, size: 4MB
[esp32] [INF] Image index: 0, Swap type: test
[esp32] [INF] Loading image 0 - slot 0 from flash, area id: 1
[esp32] [INF] Application start=40084A50h
[esp32] [INF] DRAM segment: paddr=000034D4h, vaddr=3FFB0000h, size=00BB0h (  2992) load
[esp32] [INF] IRAM segment: paddr=00004084h, vaddr=40080000h, size=054FCh ( 21756) load
[esp32] [INF] DROM segment: paddr=00010040h, vaddr=3F400040h, size=03494h ( 13460) map
[esp32] [INF] IROM segment: paddr=00020000h, vaddr=400D0000h, size=15230h ( 86576) map
*** Booting Zephyr OS build zephyr-v3.5.0 ***
Hello World! esp32_devkitc_wroom
```

Looks not bad until now.

I now wonder how to perform OTA because I don't understand the mcuboot integration. For example after adding mcuboot shell command support I get the following:

```
uart:~$ mcuboot 
swap type: test
confirmed: 1
primary area (1):
  version: 0.0.0+0
  image size: 152080
  magic: unset
  swap type: none
  copy done: unset
  image ok: unset
failed to read secondary area (2) header: -5
```

I have added few lines of code (`flash_img_init` and `flash_img_buffered_write`) to flash fix data (a 16 bytes array with values from 0 to 15) to the second slot and then verified with the shell it was properly done (`flash read` at address 0x00110000). This is working.

But at the end I added the following in the code, the function succeeds (it returns 0) but there is not effect when the board reboots:

```
boot_request_upgrade(BOOT_UPGRADE_TEST); //Checked, it returns 0
```

While rebooting the device I would expect to have a failure due to invalid data in the second slot.

It seems that the update slot is not used at all. So my question is: anybody succeeded to do OTA on ESP32, with swap or scratch of the partitions ? What is globally the status of OTA support on Espressif devices ?

https://github.com/zephyrproject-rtos/zephyr/issues/29394 indicates mcuboot is supported. Is it only booting ? @sylvioalves ?

Thanks
Joel

**Comment by rftafas** [comment]
ESP32 REV1 is not supported. May or may not be related to your problem, though.

**Reply by joelguittet** [reply]
You mean the chip revision ?
I don't know the difference to be honest but sound it's related to hardware while mcuboot ota is pure software as soon as the flash is accessible I think.
Will check if I have another board by the way.

**Reply by rftafas** [reply]
Yes, I mean chip revision. In HAL5 we will fix the issue that allows using this device, so I suggest at least getting a Rev3 device.

As for MCUboot, it should work on both ports (Espressif and Zephyr). Other than diving into the docs, you could try hawkbit or the updatehub example and check the internals for better practical examples.

**Reply by joelguittet** [reply]
Ok, will look for another board on my side definitely to see if I have such a revision.
Is esp32s3 supported maybe (whatever the revision) ? I have one of these somewhere :-)

I know a bit the details, I m the author of https://github.com/joelguittet/mender-mcu-client ;-)

**Reply by joelguittet** [reply]
@rftafas I was not able to find a rev3 here. Do you know what has been fixed exactly so that I can maybe have a better understanding of the issue ? Why isn't it ported for rev1 for example ?
Do you know how it is possible to purchase a board with a rev3 chip ?

**Reply by rftafas** [reply]
> Do you know what has been fixed exactly so that I can maybe have a better understanding of the issue ?

First, stop and PTAL: https://github.com/zephyrproject-rtos/zephyr/issues/51622 - Besides the security, lots of software workaround to make psram/flash/XIP stable are missing. Also, changes in rom functions.

> Why isn't it ported for rev1 for example ?

REV1 is obsolete.

> Do you know how it is possible to purchase a board with a rev3 chip ?

I think unless you go to some obscure used devices, everything new should be Rev3

**Reply by joelguittet** [reply]
I bought the board about 2 years ago only.... :-(

**Comment by naNEQ** [comment]
I have successfully used ESP32-S3 with MCUBoot, where a new FW image is flashed in the secondary partition over a wireless interface, the partition is marked as requested to update in test mode, exactly as you describe you did, and then at next boot the MCUBoot takes care of the swapping.

I'm not entirely sure I understand your problem, but if I get it right, you attempted to right a dummy image at the second slot? Perhaps I'm wrong, but I think this will not work. If I remember correctly, a proper image contains certain headers that the MCUBoot checks at boot time, and if not satisfied, the swap process is canceled. I recommend you to try with a proper image.

**Reply by surendersampath** [reply]
@naNEQ That sounds interesting. Do you have an example to try on the esp32-s3? I'd be keen.

**Comment by MohamedAbdElAziz18** [comment]
I have a problem with slots swapping on ESP32 Rev3.1, it only overrides slot 0 with slot1 so if the image is not OK it can't rollback, i'm trying with the example 'samples/subsys/mgmt/mcumgr/smp_svr'. I didn't change the configurations nor the code.
this how i use mcumgr cmdline:
```
(.venv) ubuntu-desktop:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ sudo mcumgr --conntype ble --connstring "peer_name=MyZephyrDevice" image list
Images:
 image=0 slot=0
    version: 0.0.2
    bootable: true
    flags: active confirmed
    hash: b8b5cfd9359ae4ccf3598145ab8dada4408fc5fc471cdfcce614680451994c6c
 image=0 slot=1
    version: 0.0.2
    bootable: true
    flags: 
    hash: 78eacb28087ce31356663b1ba8ab94a2da8dd817a511b7991f86aa710fd2e61f
Split status: N/A (0)
(.venv) ubuntu-desktop:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ sudo mcumgr --conntype ble --connstring "peer_name=MyZephyrDevice" image test 78eacb28087ce31356663b1ba8ab94a2da8dd817a511b7991f86aa710fd2e61f
Images:
 image=0 slot=0
    version: 0.0.2
    bootable: true
    flags: active confirmed
    hash: b8b5cfd9359ae4ccf3598145ab8dada4408fc5fc471cdfcce614680451994c6c
 image=0 slot=1
    version: 0.0.2
    bootable: true
    flags: pending
    hash: 78eacb28087ce31356663b1ba8ab94a2da8dd817a511b7991f86aa710fd2e61f
Split status: N/A (0)
(.venv) mohamedabdelaziz@crewteq-ubuntu-desktop:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ sudo mcumgr --conntype ble --connstring "peer_name=MyZephyrDevice" reset
Done
(.venv) ubuntu-desktop:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ sudo mcumgr --conntype ble --connstring "peer_name=MyZephyrDevice" image list
Images:
 image=0 slot=0
    version: 0.0.2
    bootable: true
    flags: active confirmed
    hash: 78eacb28087ce31356663b1ba8ab94a2da8dd817a511b7991f86aa710fd2e61f
Split status: N/A (0)
```
after reset i got only the new image and the old image has been overriden.
